retry/retry.go

```diff
@@ -0,0 +1,281 @@
+// Copyright (c) HashiCorp, Inc.
+// SPDX-License-Identifier: MPL-2.0
```

Comment:
We can't distribute this file under Apache 2, AFAIK, since MPL2 requires us to keep the original license. But keeping the MPL2 fails our CI, since it requires Apache2 on all files... do we have a solution for this?